### PR TITLE
Temporarily skip compaction for locked intervals

### PR DESCRIPTION
**Pull request by kfaraz**
Compaction Tasks, both auto and manual, try to acquire locks on the Datasource Intervals which they need to compact. If a higher priority ingestion task is in progress for an overlapping interval, the compaction task waits until it can acquire a lock. This can lead to the following potential issues:
- Due to poor configuration of `skipOffsetFromLatest`, compaction can get stuck for long periods of time.
- Once the lock is released by the ingestion task, it is possible that some new segments are published (and/or removed). This invalidates the segment spec in the compaction task, causing it to fail. Such failures are unnecessary and often difficult to debug.

### Resolution
Rather than waiting on already locked intervals, we can proceed to the compaction of other intervals (these would be older intervals, if the compaction policy is Newest First). When the locked intervals are freed up, subsequent compaction runs can submit compaction tasks for them.

In every compaction run (invocation of`CompactSegments.run()`)
- the Coordinator makes a call to the Overlord API `/lockedIntervals`
- the Overlord returns a list of intervals locked by each currently running task, using the 
  in-memory state of the `TaskLockbox`
 - the Coordinator then skips the locked intervals while submitting compaction tasks
 - For simplicity, Segment Locks are treated the same as Time Chunk Locks i.e. if a Segment in an interval
 is locked by a higher priority, the whole interval is skipped while submitting compaction tasks

### Code Changes:
 * Add Overlord REST endpoint `/druid/indexer/v1/lockedIntervals`
 * Use the above API in `CompactSegments.run()` to skip locked intervals
 * Add config `druid.coordinator.compaction.skipLockedIntervals` with default value as true

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java [hunk 1]
```diff
@@ -674,6 +675,64 @@ public TaskLock apply(TaskLockPosse taskLockPosse)
...
+    // Take a lock and populate the maps
+    giant.lock();
+    try {
+      running.forEach(
+          (datasource, datasourceLocks) -> datasourceLocks.forEach(
+              (startTime, startTimeLocks) -> startTimeLocks.forEach(
+                  (interval, taskLockPosses) -> taskLockPosses.forEach(
+                      taskLockPosse -> taskLockPosse.taskIds.forEach(taskId -> {
+                        // Do not proceed if the lock is revoked
+                        if (taskLockPosse.getTaskLock().isRevoked()) {
```
Comment (abhishekagarwal87) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:698]: what is the rationale behind this?

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java [hunk 2]
```diff
@@ -154,6 +160,17 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
...
+            taskToLockedIntervals
+        );
+        taskToLockedIntervals.forEach(
+            (taskId, datasourceIntervals) -> compactionTaskIntervals
+                .computeIfAbsent(datasourceIntervals.getDatasource(), ds -> new ArrayList<>())
+                .addAll(datasourceIntervals.getIntervals())
+        );
+
         final CompactionSegmentIterator iterator =
             policy.reset(compactionConfigs, dataSources, compactionTaskIntervals);
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:175]: maybe `compactionTaskIntervals` needs to be called something else now since it can also include intervals for which there is no lock.

File: docs/operations/api-reference.md [hunk 3]
```diff
@@ -614,6 +614,11 @@ Retrieve information about the segments of a task.
 
 Retrieve a [task completion report](../ingestion/tasks.md#task-reports) for a task. Only works for completed tasks.
 
+* `/druid/indexer/v1/lockedIntervals`
+
+Retrieve the list of Intervals locked by currently running ingestion/compaction tasks. The response contains a Map from
+Task IDs to the list of Intervals locked by the respective Tasks.
```
Comment (abhishekagarwal87) [inline docs/operations/api-reference.md:620]: why do we need task ids? can we just return a map of `datasource --> intervals`?

Comment (kfaraz) [reply docs/operations/api-reference.md:620]: In `CompactSegments.run()`, some currently running compaction tasks can get cancelled if their spec is out of date. We use the taskId to identify the intervals for these tasks getting cancelled so that we don't consider those intervals as locked.

This is the only case where taskId is useful. Otherwise, datasource would be sufficient.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java (hunk 2 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:175]: Yeah, should we just call it `skipIntervals`?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java (hunk 1 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:698]: A lock that is revoked is effectively not locking any interval. So we don't need to consider revoked locks while looking for locked intervals.

Revoked locks are kept in the TaskLockbox to notify that those locks are revoked to the callers when they acquire the same locks again.

File: docs/operations/api-reference.md (hunk 3 above)
Comment (kfaraz) [reply docs/operations/api-reference.md:620]: See here:
https://github.com/apache/druid/blob/5eb91aba52e22d9c72fd36f3994dc9b103ff5a69/server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java#L147

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java [hunk 4]
```diff
@@ -227,6 +227,34 @@ public Response isLeader()
     }
   }
 
+  @GET
+  @Path("/lockedIntervals")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getTaskLockedIntervals(@Context HttpServletRequest request)
+  {
+    // Perform authorization check
+    final ResourceAction resourceAction = new ResourceAction(
```
Comment (maytasm) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:236]: Any particular this API needs a finer grained access control rather than using the StateResourceFilter ?
The StateResourceFilter is already used for API like /taskStatus which I think is similar in access control to this API.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java [hunk 5]
```diff
@@ -227,6 +227,34 @@ public Response isLeader()
...
+                  .entity(StringUtils.format("Access-Check-Result: %s", authResult.toString()))
+                  .build()
+      );
+    }
+
+    // Build the response
+    final LockedIntervalsResponse response = new LockedIntervalsResponse(
+        taskStorageQueryAdapter.getLockedIntervals()
+    );
+    log.warn("Found Intervals: %s", response.getLockedIntervals());
```
Comment (maytasm) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:254]: Why is this a WARN log?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java [hunk 6]
```diff
@@ -227,6 +227,34 @@ public Response isLeader()
...
+    if (!authResult.isAllowed()) {
+      throw new WebApplicationException(
+          Response.status(Response.Status.FORBIDDEN)
+                  .entity(StringUtils.format("Access-Check-Result: %s", authResult.toString()))
+                  .build()
+      );
+    }
+
+    // Build the response
+    final LockedIntervalsResponse response = new LockedIntervalsResponse(
```
Comment (maytasm) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:251]: Can the API just return Map<String, DatasourceIntervals> (removing the need for another class)?

Comment (maytasm) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:251]: Response.ok(taskStorageQueryAdapter.getLockedIntervals()).build()

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java [hunk 7]
```diff
@@ -139,6 +142,9 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
                          compactionTaskQuery.getGranularitySpec().getSegmentGranularity(),
                          configuredSegmentGranularity);
                 indexingServiceClient.cancelTask(status.getId());
+
+                // Remove this from the locked intervals
+                taskToLockedIntervals.remove(status.getId());
```
Comment (maytasm) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:147]: As mentioned earlier, this is not really needed. This would be the same as taskToLockedIntervals not containing lock for compaction task in the first place.

**Review by maytasm — CHANGES_REQUESTED** [review]
- I think what we can do here is use the taskPriority set for the auto compaction of each dataSource. The lockedIntervals API can then only returns lockedIntervals of task with priority greater than the taskPriority set for the auto compaction. For example, if the task that is holding the lock has priority lower than compaction task to be scheduled, we should schedule it as it will revoke the lock of the current task (and not skip it). The priority of compaction task scheduled by auto compaction can also be set by the user (taskPriority field) and revoke locks of batch/stream ingestion task. 
- Can you add integration tests too? Maybe a simple one for the lockedIntervals API and a simple one to check that auto compaction skipped locked intervals.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java [hunk 8]
```diff
@@ -674,6 +675,64 @@ public TaskLock apply(TaskLockPosse taskLockPosse)
     }
   }
 
+  /**
+   * Gets a Map containing intervals locked by active tasks. Intervals locked
+   * by revoked TaskLocks are not included in the returned Map.
+   *
+   * @return Map from Task Id to locked intervals.
+   */
+  public Map<String, DatasourceIntervals> getLockedIntervals()
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:684]: `getLockedInterval()` seems a misnomer because this method returns the intervals of segment locks as well, but they don't lock intervals. I don't have a better suggestion though..

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java [hunk 9]
```diff
@@ -154,6 +160,17 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
           }
         }
 
+        // Skip all the locked intervals
+        LOG.debug(
+            "Skipping the following intervals for Compaction as they are currently locked: %s",
+            taskToLockedIntervals
+        );
+        taskToLockedIntervals.forEach(
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java:168]: I think it should behave differently depending on what `lockGranularity` is used. If both the compaction task to run and the task that is already running use the segment lock, the compaction task can safely run. Otherwise, the entire locked interval should be skipped as what this code does.

File: docs/operations/api-reference.md [hunk 10]
```diff
@@ -614,6 +614,11 @@ Retrieve information about the segments of a task.
 
 Retrieve a [task completion report](../ingestion/tasks.md#task-reports) for a task. Only works for completed tasks.
 
+* `/druid/indexer/v1/lockedIntervals`
```
Comment (jihoonson) [inline docs/operations/api-reference.md:617]: Is this supposed to be called by users? What is the use case?

Comment (maytasm) [reply docs/operations/api-reference.md:617]: Actually, instead of exposing this API, maybe a better thing we need is how to communicate to the user that compaction skips an intervals because of some task holding lock. Like what intervals was skipped for which datasource and what was the task that holds the lock.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java (hunk 8 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:684]: The response would contain a List of `Interval` objects (either locked by TimeChunkLocks or SegmentLocks). So I guess `getLockedIntervals` should be fine for now?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java (hunk 4 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java]: Fixed.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java (hunk 5 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java]: Removed.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java (hunk 6 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java]: Fixed.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java (hunk 7 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java]: Removed.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java (hunk 9 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java]: For simplicity, we are treating Segment Locks the same as Time Chunk Locks i.e. the whole interval would be skipped while submitting compaction tasks even if there is just one Segment in that interval that is locked by a higher priority task.

Added this as a javadoc comment here:
https://github.com/apache/druid/blob/a8e30c3c8c9c2b58b0615f3f50d863c334cadd75/server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java#L233

https://github.com/apache/druid/blob/a8e30c3c8c9c2b58b0615f3f50d863c334cadd75/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java#L690

File: docs/operations/api-reference.md (hunk 10 above)
Comment (kfaraz) [reply docs/operations/api-reference.md]: For now, we are just logging the intervals that were skipped due to locks.
The part about notifying the user about skipped intervals will most likely be done through Task Reports in a follow up PR.

Comment (kfaraz) [reply docs/operations/api-reference.md]: > Is this supposed to be called by users? What is the use case?
Removed changes to `api-reference.md` as this API is for internal use (between Coordinator and Overlord) only.

**Comment by kfaraz** [comment]
Fixed lockedIntervals API to use taskPriority
Added integration tests.

**Review by maytasm — CHANGES_REQUESTED** [review]
Minor comments

**Comment by abhishekagarwal87** [comment]
Thank you for the PR, @kfaraz. will there be a way to disable this code-path in a running druid cluster (either dynamically or by restarting services)

**Comment by rohangarg** [comment]
@kfaraz : Thank a lot for the PR - I have a couple of doubts regarding the new config added : 
1. does the config `druid.coordinator.compaction.skipLockedIntervals` indicate that the segments which are locked during compaction are missed forever? or is it just for that run?
2. what would be the user impact of this config? for example, if I'm a user when should I enable/disable this config and what will be its benefits/problems?

**Comment by kfaraz** [comment]
>     1. does the config `druid.coordinator.compaction.skipLockedIntervals` indicate that the segments which are locked during compaction are missed forever? or is it just for that run?

@rohangarg , the locked intervals would be skipped only for that run, they would be retried in the next run
 
>     2. what would be the user impact of this config? for example, if I'm a user when should I enable/disable this config and what will be its benefits/problems?

Normally, the user would not have to specify any value for this config. The default value of true would suffice. The config has been provided to be able to disable the feature of skipping locked intervals in case of a bug.

**Comment by maytasm** [comment]
- Can you add a test in ITAutoCompactionLockContentionTest that verify that setting the new flag `druid.coordinator.compaction.skipLockedIntervals` to false will actually disable this new feature.
- Actually if getLockedIntervalsToSkip does return lock of Compaction task, then you will need https://github.com/apache/druid/pull/11190#discussion_r625757970
(as you mentioned, in CompactSegments.run(), some currently running compaction tasks can get cancelled and hence we have to return the lock)

**Comment by rohangarg** [comment]
>Normally, the user would not have to specify any value for this config. The default value of true would suffice. The config has been provided to be able to disable the feature of skipping locked intervals in case of a bug.
Ok but given that the feature/change-set is not too big, I'm not sure whether adding a config would help. Is it possible to know some scenarios or unknowns where this could fail and might be hard to cover them in tests? 

Apologies but one more question I had was that if this flag is turned on, will it always ensure that there is no locking conflict between compaction task and other tasks? or is it still possible that compaction task might get stuck due to unavailable lock? my initial guess is that there is still a chance that compaction might get stuck, so confirming? :)

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java [hunk 11]
```diff
@@ -710,8 +710,8 @@ public TaskLock apply(TaskLockPosse taskLockPosse)
                             // Do not proceed if the lock is revoked
                             return;
                           } else if (taskLockPosse.getTaskLock().getPriority() == null
-                                     || taskLockPosse.getTaskLock().getPriority() <= minTaskPriority.get(datasource)) {
-                            // Do not proceed if the lock has a priority less than or equal to the minimum
+                                     || taskLockPosse.getTaskLock().getPriority() < minTaskPriority.get(datasource)) {
```
Comment (maytasm) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:713]: Why is this changed back?

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:713]: I had missed changing it the first time. 

The behaviour after this change is that intervals of tasks with equal priority will be considered locked. That is the required behaviour, right?

Comment (maytasm) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:713]: LGTM

**Review by maytasm — APPROVED** [review]
